### ISSUE TITLE
Logstash crashes fails on Master when run with jruby.jit.threshold=0

### ISSUE THREAD

**Issue by robbavey** [elastic/logstash#8030]
Running the following pipeline on master with dead letter queue and enabled and`-Djruby.jit.threshold=0` specified in `jvm.options`, Logstash crashes consistently:

```
 input { 
  generator{ 
   message => '{"test":"one"}' 
   codec => "json"                                                                                                                                                        
   count => 1000 
  }
 dead_letter_queue {
   path => "/Users/robbavey/code/logstash/data/dead_letter_queue" 
   commit_offsets => true 
  } 
 } 
 filter { 
  if ([geoip]) { 
    mutate { 
     remove_field => ["geoip"] 
     add_field => { 
      "mutated" => "true" 
      } 
    } 
   }else{ 
    mutate { 
      add_field => { 
       "geoip" => "somewhere" 
      } 
   } 
  } 
 } 
 output { elasticsearch {} }
```

with one of the following two errors:
```
[2017-08-18T15:45:22,348][FATAL][logstash.runner          ] An unexpected error occurred! {:error=>#<NoMethodError: undefined method `[]=' for #<LogStash::OutputDelegator:0x1c7a8ff4>>, :backtrace=>["/Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:522:in `block in output_batch'", "org/jruby/RubyArray.java:1734:in `each'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:529:in `block in output_batch'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:227:in `block in each'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:226:in `each'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:523:in `output_batch'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:481:in `worker_loop'", "/Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:437:in `block in start_workers'"]}
NoMethodError: undefined method `[]=' for #<LogStash::OutputDelegator:0x1c7a8ff4>
   block in output_batch at /Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:522
                    each at org/jruby/RubyArray.java:1734
   block in output_batch at /Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:529
           block in each at /Users/robbavey/code/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:227
                    each at /Users/robbavey/code/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:226
            output_batch at /Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:523
             worker_loop at /Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:481
  block in start_workers at /Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:437
```

or 

```
Unhandled Java exception: java.lang.NoSuchMethodError: 
java.lang.NoSuchMethodError: 
           block in output_batch at /Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:529
                      callDirect at org/jruby/runtime/CompiledIRBlockBody.java:145
                      callDirect at org/jruby/runtime/MixedModeIRBlockBody.java:113
                            call at org/jruby/runtime/IRBlockBody.java:71
                            call at org/jruby/runtime/Block.java:124
                            call at org/jruby/RubyProc.java:289
                          call19 at org/jruby/RubyProc.java:273
                            call at org/jruby/RubyProc$INVOKER$i$0$0$call19.gen:-1
                            call at org/jruby/internal/runtime/methods/DynamicMethod.java:204
                            call at org/jruby/internal/runtime/methods/DynamicMethod.java:200
                            call at org/jruby/runtime/callsite/CachingCallSite.java:161
               invokeOther1:call at Users/robbavey/code/logstash/logstash_minus_core/lib/logstash/util//Users/robbavey/code/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:227
                   block in each at /Users/robbavey/code/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:227
                     yieldDirect at org/jruby/runtime/CompiledIRBlockBody.java:156
                           yield at org/jruby/runtime/BlockBody.java:114
                           yield at org/jruby/runtime/Block.java:165
                            each at org/jruby/javasupport/ext/JavaLang.java:93
                            each at org/jruby/javasupport/ext/JavaUtil.java:133
                            call at org/jruby/javasupport/ext/JavaUtil$Collection$INVOKER$s$0$0$each.gen:-1
                       callBlock at org/jruby/runtime/callsite/CachingCallSite.java:139
                            call at org/jruby/runtime/callsite/CachingCallSite.java:145
               invokeOther5:each at Users/robbavey/code/logstash/logstash_minus_core/lib/logstash/util//Users/robbavey/code/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:226
                            each at /Users/robbavey/code/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:226
  RUBY$method$each$0$__VARARGS__ at Users/robbavey/code/logstash/logstash_minus_core/lib/logstash/util//Users/robbavey/code/logstash/logstash-core/lib/logstash/util/wrapped_synchronous_queue.rb:-1
                            call at org/jruby/internal/runtime/methods/CompiledIRMethod.java:77
                            call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:93
                       callBlock at org/jruby/runtime/callsite/CachingCallSite.java:77
                            call at org/jruby/runtime/callsite/CachingCallSite.java:83
                       interpret at org/jruby/ir/instructions/CallBase.java:428
                     processCall at org/jruby/ir/interpreter/InterpreterEngine.java:355
                       interpret at org/jruby/ir/interpreter/StartupInterpreterEngine.java:73
                       interpret at org/jruby/ir/interpreter/InterpreterEngine.java:83
                INTERPRET_METHOD at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:179
                            call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:165
                            call at org/jruby/internal/runtime/methods/DynamicMethod.java:200
                            call at org/jruby/runtime/callsite/CachingCallSite.java:161
      invokeOther20:output_batch at Users/robbavey/code/logstash/logstash_minus_core/lib/logstash//Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:481
                     worker_loop at /Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:481
                            call at org/jruby/internal/runtime/methods/CompiledIRMethod.java:116
                            call at org/jruby/internal/runtime/methods/MixedModeIRMethod.java:198
                            call at org/jruby/internal/runtime/methods/DynamicMethod.java:208
                    cacheAndCall at org/jruby/runtime/callsite/CachingCallSite.java:358
                            call at org/jruby/runtime/callsite/CachingCallSite.java:195
        invokeOther0:worker_loop at Users/robbavey/code/logstash/logstash_minus_core/lib/logstash//Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:437
          block in start_workers at /Users/robbavey/code/logstash/logstash-core/lib/logstash/pipeline.rb:437
                      callDirect at org/jruby/runtime/CompiledIRBlockBody.java:145
                            call at org/jruby/runtime/IRBlockBody.java:71
                            call at org/jruby/runtime/Block.java:124
                            call at org/jruby/RubyProc.java:289
                            call at org/jruby/RubyProc.java:246
                             run at org/jruby/internal/runtime/RubyRunnable.java:104
                             run at java/lang/Thread.java:748
```

Note that invoke dynamic doesn't appear to have much of an effect on the frequency of this error

**Comment by robbavey** [comment]
cc @original-brownbear

**Comment by original-brownbear** [comment]
@robbavey related to #7999 :) I'll test the fixes provided by the JRuby team for this soon :)

**Comment by robbavey** [comment]
Yeah, that looks similar - if not the same. This fails pretty much 100% of the time too...

**Comment by original-brownbear** [comment]
this should also be fixed in the next JRuby release thanks to https://github.com/jruby/jruby/issues/4739#issuecomment-323391052 :)

**Comment by original-brownbear** [comment]
fixed thanks to JRuby `9.1.13.0` :)
